packages/datadog-plugin-aws-sdk/src/index.js

```diff
@@ -0,0 +1,90 @@
+'use strict'
+
+const Tags = require('opentracing').Tags
+const analyticsSampler = require('../../dd-trace/src/analytics_sampler')
+const awsHelpers = require('./helpers')
+
+function createWrapRequest (tracer, config) {
+  config = normalizeConfig(config)
+
+  return function wrapRequest (send) {
+    return function requestWithTrace (cb) {
+      if (!this.service) return send.apply(this, arguments)
+
+      const serviceName = this.service.serviceIdentifier
+      const childOf = tracer.scope().active()
+      const baseTags = {
+        [Tags.SPAN_KIND]: 'client',
+        'span.type': 'http',
+        'service.name': config.service
+          ? `${config.service}-aws-${serviceName}`
+          : `${tracer._service}-aws-${serviceName}`,
+        'aws.operation': this.operation,
+        'aws.region': this.service.config && this.service.config.region,
+        'aws.service': this.service.api && this.service.api.className,
+        'aws.url': this.service.endpoint && this.service.endpoint.href,
+        'component': 'aws-sdk'
+      }
+
+      const span = tracer.startSpan('aws.request', {
+        childOf,
+        tags: baseTags
+      })
+
+      const boundCb = typeof cb === 'function' ? tracer.scope().bind(cb, childOf) : cb
+
+      this.on('complete', response => {
+        if (!span) return
+
+        awsHelpers.addResponseTags(span, response, serviceName, config)
+        awsHelpers.finish(span, response.error)
+      })
+
+      analyticsSampler.sample(span, config.analytics)
+
+      return tracer.scope().activate(span, () => {
+        return send.call(this, boundCb)
+      })
+    }
+  }
+}
+
+function normalizeConfig (config) {
+  const hooks = getHooks(config)
+
+  return Object.assign({}, config, {
+    hooks
+  })
+}
+
+function getHooks (config) {
+  const noop = () => {}
+  const request = (config.hooks && config.hooks.request) || noop
+
+  return { request }
+}
+
+// <2.1.35 has breaking changes for instrumentation
+// https://github.com/aws/aws-sdk-js/pull/629
+module.exports = [
+  {
+    name: 'aws-sdk',
+    versions: ['>=2.3.0'],
+    patch (AWS, tracer, config) {
+      this.wrap(AWS.Request.prototype, 'promise', createWrapRequest(tracer, config))
```

Comment:
Just to make sure, `promise` never calls `send`?

Reply:
No, not afaik from looking through the definition and through testing (the case we're worried about here is causing double patching?). The promise code is tremendously difficult to follow but this is basically the guts of it, basically our patching doesn't seem to pick up the `complete` event without wrapping `promise` specifically 
https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L779